client structure for ORAM

JosePisco opened the pull request.

Commit:
[feat] Added client side implementation

Implemented client side structure containing a stash, and encryption
key pairs. Decentralized generate_dummies() and block_change() methods
for more proper, generic ones.

Commit:
[feat] Client file code

Commit:
[feat] Finished concept of ORAM client

Client holds a csprng and a secret key for decryption. Still things to do

Commit:
[test] Added more tests to test ORAM client. Tests.

Commit:
[fix] Overall code improvement

tbrezot commented on src/btree.rs:
```diff
@@ -3,15 +3,13 @@ use crate::oram::BUCKET_SIZE;
         let mut tree = BTree {
             root: Option::None,
-            nb_blocks,
             height: nb_blocks.ilog2() as u16 + 1,
```
This conversion seems strange. Don't you want a ceiling function instead?

JosePisco replied:
No, since a complete tree has `2^x - 1` nodes and has a height of `x`. If you init BTree with a power of 2 as the number of nodes, then a ceiling function will not "plus one" to the expected value, since 2^x nodes result in an incomplete binary tree with only one node on the last level.

A floor function with + 1 is in my opinion the best thing to do.

tbrezot commented on src/btree.rs:
```diff
@@ -3,15 +3,13 @@ use crate::oram::BUCKET_SIZE;
 }
 
 impl BTree {
     pub fn init_new(dummies: &mut Vec<DataItem>, nb_blocks: usize) -> BTree {
```
The idiomatic way of passing reference on a sequence is using a slice (in order to abstract the backing structure and allow more flexibility):
```suggestion
    pub fn init_new(dummies: &mut [DataItem], nb_blocks: usize) -> BTree {
```

JosePisco replied:
But there would be no methode to remove element as in line 55 ?
```rs
node.set_bucket_element(dummies.remove(j), i);
```

tbrezot replied:
True. I didn't see that. This comment still holds for the other functions. For this one, you may want to use another structure instead (linked list?) to avoid the `O(n)` cost in the worst case to call `.remove()`.

To avoid going trough your sequence of items several times, you could go through once and store the indices of (or pointers to) the elements with a matching path.

JosePisco replied:
Maybe switching to a linked list would be better ? In the meantime I think of using `swap_remove()` method in $O(1)$

tbrezot replied:
Great, I didn't know this method. I was about to propose you to set the path to `u16::MAX` or something like that to avoid considering it in a later time, but `swap_remove()` is much better since it reduces the list size.

tbrezot commented on src/btree.rs:
```diff
@@ -3,15 +3,13 @@ use crate::oram::BUCKET_SIZE;
 }
 
 impl BTree {
     pub fn init_new(dummies: &mut Vec<DataItem>, nb_blocks: usize) -> BTree {
```
Why don't you initialize your tree with the actual data? It seems that you are storing one block per node (from the use of log). If you pass the data directly you could then omit `nb_blocks`.

tbrezot replied:
This comment may be off since it is unclear to me what is a `block`. After a complete review, it seems `dummy` and `block` are both `DataItem`s. I suggest you review your naming. It is quite hard for me to follow you sometimes.

JosePisco replied:
I think removing the concept of block is the best since it could be misunderstood with symmetric encryption blocks.

JosePisco replied:
Done in commit eb2f5b258074b714751eda363bb6f78e218a591a

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        let key = SymmetricKey::new(&mut csprng);
+
+        ClientORAM {
+            // Empty stash at initialization.
```
redundant comment
```suggestion
            // Empty stash at initialization.
```

JosePisco replied:
I'm not sure how can it be redundant ?
Maybe precise that it is as described in `https://eprint.iacr.org/2013/280` ?

tbrezot replied:
The line of code below this comment is explicit enough. It reads "initialise `ClientOram.stash` to an empty `Vec`".

tbrezot replied:
Btw, you are not supposed to use upper case acronyms in Rust: `ORAM` -> `Oram`.
https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

JosePisco replied:
I get that, but it feels odd, doesn't it ?

tbrezot replied:
You will get used to it :wink:

JosePisco replied:
Done in 1252d24c773e57796a20eb0f21a865a7c2031bd2

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+
+    pub fn gen_key(&mut self) {
+        self.key = SymmetricKey::new(&mut self.csprng);
+    }
```
Why do you need a `key_gen()` _and_ an initialization of the key in `new()`?

JosePisco replied:
Indeed, this is a blunder. Thank you for pointing this out.

Correction in 87d5f5d8927a85bd703eb1fb7e4c57a139c1417d

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        self.key = SymmetricKey::new(&mut self.csprng);
+    }
+
+    pub fn generate_dummies(
```
I am not a fan of this naming: `dummies` does not give any information on _what_ is generated.

JosePisco replied:
You are right, I shall comment better on it and rename it to probably `data_item` since at initialization, if the values are provided by the client, they might be dummies as well as encrypted data.

JosePisco replied:
Done in a7dee6de99c73da1f81fc66b650302869938a62d

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        if !(nb_dummies + 1).is_power_of_two() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Number of blocks shall be power of 2 minus one".to_string(),
```
Vocabulary is misleading here. Maybe consider renaming `DataItem` to `Block`.

tbrezot replied:
I would not enforce this condition here though. There seems to be no inherent constraint on `generate_dummy`. My guess is that this comes from something else. Enforce this condition where it is needed and give a rational if the context is not clear why.

JosePisco replied:
I see your point, since the client should be able to generate dummies **and** insert encrypted data, it makes sense not to restrict the number of dummy items to generate.

Done in f96fcb0d813824aa61c447b1d01f2bdef548892b

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+
+        // Number of leaves is 2^(l-1) if number of blocks is 2^l - 1.
+        let nb_leaves = (nb_dummies + 1) / 2;
+        let mut dummies = Vec::new();
```
Allocate the correct size here:
```suggestion
        let mut dummies = Vec::with_capacity(nb_dummies);
```

JosePisco replied:
Shouldn't it be
```rs
let mut dummies = Vec::with_capacity(nb_dummies * BUCKET_SIZE);
```
?

tbrezot replied:
Nope, your definition of `nb_dummies` is wrong. It actually counts the nodes (the size of `dummies` that you are declaring should be `nb_dummies`, otherwise this is obfuscating and you should rename one or the other).

JosePisco replied:
Commit aa8979d36c2b5943f9e657417a65ca8e58b92eca

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        let cipher = Aes256Gcm::new(&self.key);
+
+        // FIXME - encrypt fixed dummy value instead of encrypting randoms.
+        for _ in 0..nb_dummies * BUCKET_SIZE {
```
It seems that `nb_dummies` is not the number of dummies... but rather a  number of nodes to fill.

JosePisco replied:
Right, expression mistake. Removed BUCKET_SIZE so it generates the given number of dummies.

Done in 83d3bc7feb9e65263d9b4b176fb11b767a95774d

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+
+        // FIXME - encrypt fixed dummy value instead of encrypting randoms.
+        for _ in 0..nb_dummies * BUCKET_SIZE {
+            // Generate new random dummy data.
```
redundant comment
```suggestion
```

JosePisco replied:
I don't think it is too much to explain the reasoning here.

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+                    ErrorKind::Interrupted,
+                    Result::unwrap_err(ciphertext_res).to_string(),
+                ));
+            }
```
The comment is wrong imho.

You can use a `map_err` to convert the error and `?` to unwrap it.
```suggestion
            let encrypted_data = cipher.encrypt(&nonce, &dummy_data, None).map_err(|e|
                Error::new(ErrorKind::Interrupted, e.to_string()))?;
```

JosePisco replied:
How is this comment wrong ? If you try to GCM decrypt random data, the concatenated MAC will not verify random data as if it had been encrypted ?

JosePisco replied:
Concerning errors, I didn't know you could handle them like that! So much better :)

JosePisco replied:
Done in 52dd2d82e4d6d48398c8fb028095f40120e6de74

Tho, I'm curious to know your position on the comment being inaccurate.

tbrezot replied:
It is actually misleading rather than wrong. You encrypt 0s in order to have a seemingly valid value. The server won't be able to check the MAC. You could even use a wrong mac to discriminate dummies from valid encrypted values.

Or maybe I am missing something here.

JosePisco replied:
First thing is we should not encrypt a vector of 0s since it is used internally in GCM to compute the authentication key that holds for the MAC

Second thing, discriminating dummies could be done by checking if the MAC is invalid but if a malicious third party interferes with the communications and changes the ciphertexts, we wouldn't be able to tell.

How to discriminate them is matter of discussion more than code reviewing I think.

tbrezot replied:
I agree for your second point.

For the first, I don't know that. I will check. Anyway you still can encrypt 1s and my point still stands. Anyway, this is setup and generating too much random shouldn't be a problem so do as you please. Encrypting random just sounds redundant to me.

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        for _ in 0..nb_dummies * BUCKET_SIZE {
+            // Generate new random dummy data.
+            let mut dummy_data = vec![0; block_size];
+            self.csprng.fill_bytes(&mut dummy_data);
```
You don't need to encrypt random data. The ciphertext is already random-like.
```suggestion
```

JosePisco replied:
Regarding previous conversation, yes you need to encrypt it to provide correct and verifiable MAC via GCM.

tbrezot replied:
I disagree. What you need is a seemingly valid ciphertext:
- You could use true random instead. Anyway the server won't be able to check the MAC. This saves encrypting random.
- You could need a valid MAC. Therefore you could generate a real ciphertext. However you don't need to encrypt a random, encrypting 0s is enough since your nonce is random and AES guarantees indistinguishably from true encrypted random. This saves generating 32-bytes randoms.

JosePisco replied:
I close this conversation since the same topic is discussed above.

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+            // not need to be generated at random.
+            let path = self.csprng.gen_range(0..nb_leaves);
+
+            dummies.push(DataItem::new(encrypted_dummy, path as u16));
```
You will generate a `u16` if you convert `nb_leaves` to `u16`. But I think it is okay to use a `usize` for the path (I feel like a `u16` may be too small). Plus, as is, you may have an integer overflow during this cast (constrain `nb_dummies` so that it can't happen or add a check on its size).
```suggestion
            let path = self.csprng.gen_range(0..nb_leaves);

            dummies.push(DataItem::new(encrypted_dummy, path));
```

JosePisco replied:
I think you are right, I expected no more than 2^16 levels hence u16 for the height, but path being a u16 is wrong  and will overflow (thus panic) when you insert more than 2^17 items.

JosePisco replied:
Done in 3be5899d52f3417bdf3412511396b7a280794661

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        block_ids: Vec<usize>,
+        max_path: usize,
+    ) -> Result<(), CryptoCoreError> {
+        self.gen_key();
```
It seems you are throwing the previous key away here.

JosePisco replied:
Yup, as discussed previously this has been corrected in 87d5f5d8927a85bd703eb1fb7e4c57a139c1417d

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        let cipher = Aes256Gcm::new(&self.key);
+
+        let mut i = 0;
+        while i < blocks.len() {
```
Come on... just use a loop.

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+                return Err(Result::unwrap_err(ciphertext_res));
+            }
+
+            // Change element data to plaintext.
```
Wrong comment. I wouldn't add any here anyway. What you are doing is pretty obvious.
```suggestion
```

JosePisco replied:
Small mistake on the comment indeed. Disagreed still on removing comments, it's probably obvious but it's good to think of people looking at it without context and things like that might help them. At worst, they'll say "oh yes of course".

tbrezot replied:
Comment is still wrong (maybe you didn't update it yet, but unresolving this conv').

JosePisco replied:
What do you mean by that ?

tbrezot replied:
My bad I was looking at the wrong version of the file.

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+             */
+            if block_ids.contains(&i) {
+                blocks[i].set_path(self.csprng.gen_range(0..max_path as u16))
+            }
```
You could use a more explicit name, which would avoid the need for a comment. Plus, constrain your path in the signature to avoid the need for this dangerous cast.
```suggestion
            if blocks_to_move.contains(&block_id) {
                blocks[block_id].set_path(self.csprng.gen_range(0..max_path))
            }
```

JosePisco replied:
True; changed for `changed_items_idx`

Done in eb2f5b258074b714751eda363bb6f78e218a591a

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+        &self,
+        blocks: &mut Vec<DataItem>,
+    ) -> Result<(), CryptoCoreError> {
+        let cipher = Aes256Gcm::new(&self.key);
```
I really don't understand how this can work since you lose the first key after calling `encrypt_blocks` twice.

JosePisco replied:
Yup, as discussed previously this has been corrected in https://github.com/Cosmian/oram/commit/87d5f5d8927a85bd703eb1fb7e4c57a139c1417d

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+    ) -> Result<(), CryptoCoreError> {
+        let cipher = Aes256Gcm::new(&self.key);
+
+        let mut i = 0;
+        while i < blocks.len() {
+            // Edge-case where dummies left cells uninitialized.
+            if blocks[i].data().is_empty() {
+                i += 1;
+                continue;
+            }
```
Loop on an iterator instead.
```suggestion
        for block in blocks {
            // Edge-case where dummies left cells uninitialized.
            if block.data().is_empty() {
                continue;
            }
```

JosePisco replied:
But doesn't this method make a copy of the element we are working with ?
If I change block, do I change a copy of it or is it passed by reference ?

If the latter, then this method sounds better.

JosePisco replied:
fixed for an enumerator in commit 6b498d41c1a8a025a0d7f2e6f7d04a456619a1b8

tbrezot commented on src/client.rs:
```diff
@@ -0,0 +1,162 @@
+                continue;
+            }
+
+            let nonce_res = Nonce::try_from_slice(
+                &blocks[i].data()[..Aes256Gcm::NONCE_LENGTH],
+            );
+            if nonce_res.is_err() {
+                return Err(Result::unwrap_err(nonce_res));
+            }
+
+            let nonce = nonce_res.unwrap();
+            let plaintext_res = cipher.decrypt(
+                &nonce,
+                &blocks[i].data()[Aes256Gcm::NONCE_LENGTH..],
+                Option::None,
+            );
+
+            if plaintext_res.is_err() {
+                return Err(Result::unwrap_err(plaintext_res));
+            }
```
I am pretty sure there is a function in `crypto_core` that performs exactly this.

tbrezot replied:
I checked and this is not true anymore for the latest release of CryptoCore...

tbrezot commented on src/oram_tests.rs:
Move the dedicated test files in `tests/` in the root of your repository. Test modules in `src` are meant for unit tests.

JosePisco replied:
Correct, my bad.

Should I differentiate unit tests and testsuites ?
Unit tests should be in the file of the method or in a separate one ?

tbrezot replied:
You can add your unit tests at the end of the corresponding file:
```rust

pub fn my_func(...) {...}

#[test]
mod tests {
    use super::*;
    
    fn test_my_func() {...}
}
```

tbrezot reviewed (commented):
It seems this MR is based on an anterior commit that has not passed any review. As such, I cannot comment unchanged lines from this anterior commit.

About `complete_tree`:
- it seems the definition of `level` is not easy to use, maybe consider changing it to `level + 1` or deducing it from `log(path)`.
- I think the performance can be greatly improved. Indeed `remove`'s worst case is `O(n)` since this is a linear array. Plus it seems that the inner loop is not needed. Instead, use a sorted (on `dummy.path`) stack of the dummies (`O(log(n))` vs `O(n)`).
- the comment talking about `leaves` is misleading since you are performing a depth-first traversal (the rightmost leaf will be filled first, but the leftmost leaf will be filled *after* the right node of depth 1)
- after reading your `generate_dummies`, I wonder if you cannot just fill the buckets with ciphertexts of 0s and setting the path of the `DataItem` accordingly.

Commit:
[fix] Better naming of item elements; removed ambiguous block name

Commit:
[fix] /!\ Removed gen_key for new encryptions

Old code generated a new key every encrypt() operation. It would make
other values in the tree undecryptable. Client has one key and stores
a cipher object now instead of his secret key.

Commit:
[fix] Better naming for dummy_items and data_items

Commit:
[fix] Removed size restriction in dummy items generation

Commit:
[fix] generate_dummy_item() generates given number of dummy items

Commit:
[fix] Better error handling with ? operator

Commit:
[fix] Changed `path` related variables to usize type instead of u16

Making it u16 was an error and this could overflow (or panic) when one
inserted more than 2^17 elements.

Commit:
[fix] Overall naming and comments improvement

Commit:
[feat] Resize dummy vector accroding to its capacity

Commit:
[chg] Changed while loops for enumerator

Commit:
[feat] Removed FIXMEs and added encrypt/decrypt stash client side
